File: leetcode/algorithms/linked-list/reverse-linked-list-ii.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Linked List II"
*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode trailNode = null,leadNode=null,currNode=head,severedTail=null,severedHead=null;
        int counter = 0;
        while(currNode!=null){
            ++counter;
            if(counter==m-1){
                severedTail=currNode;
            }
            if(counter>=m && counter<=n){
                leadNode = currNode.next;
                currNode.next=trailNode;
                if(trailNode==null){
                    severedHead=currNode;
                }
                trailNode=currNode;
                currNode=leadNode;
            } else {
                currNode=currNode.next;
            }
            if(counter==n){
                if(severedTail!=null){
                    severedTail.next=trailNode;
                } else {
                    head=trailNode;
                }
                severedHead.next=currNode;
                break;
            }
        }
        return head;
        
    }
}

File: leetcode/algorithms/linked-list/reverse-linked-list-ii.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Linked List II"
*/


/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseBetween(head: ListNode | null, m: number, n: number): ListNode | null {
    let counter: number = 1;
    if(head===null){
        return null;
    }
    let p: ListNode | null = head, t: ListNode | null = null;
    while(counter<m){
        if(counter===m-1){
            t=p;
        }
        p = p!.next;
        ++counter;
    }
    let h = p;
    let l: ListNode | null = null, q = p;
    while(counter<=n){
        q = p!.next;
        p!.next = l;
        l = p;
        p = q;
        ++counter;
    }
    
    if(t===null){
        head = l;
    } else {
        t!.next = l;
    }
    h!.next = p;
    
    return head;
        
};